tfx_addons/sampler/executor.py

```diff
@@ -0,0 +1,247 @@
+  if parsed.features.feature[label].int64_list.value:
+    val = parsed.features.feature[label].int64_list.value
+    if len(val) > 0:
+      class_label = val[0]
+  else:
+    val = parsed.features.feature[label].bytes_list.value
+    if len(val) > 0:
+      class_label = val[0].decode()
+  return (class_label, parsed)
+
+
+def sample_data(_, val, undersample=True, side=0):
+  if undersample:
+    random_sample_data = random.sample(val, side)
+  else:
+    random_sample_data = random.choices(val, k=side)
+
+  for item in random_sample_data:
+    yield item
+
+
+def filter_null(item, keep_null=False, null_vals=None):
+  """Function that filters all of the null labels (and any optional labels
+  from the dataset at large. Returns either the null-labeled values or the
+  non-null-labeled values, depending on the value of keep_null."""
+
+  if item[0] == 0:
+    keep = True
+  else:
+    keep = bool(item[0])
+
+  if null_vals and str(item[0]) in null_vals and keep:
+    keep = False
+  keep ^= keep_null
+
+  if keep:
+    return item
+  else:
+    return None
+
+
+def sample(p, uri, label, shards, keep_classes, output_dir, undersample):
+  """Function that actually samples the given split.
+
+  Args:
+    uri: The input uri for the specific split of the input example artifact.
+    label: The name of the column containing class names to undersample by.
+    shards: The number of files that each undersampled split should
+    contain. Default 0 is Beam's tfrecordio function's default.
+    keep_classes: A list determining which classes that we should
+    not undersample. Defaults to None.
+    output_dir: The output directory for the split of the output artifact.
+  Returns:
+    None
+  """
+
+  data = read_tfexamples(p, uri, label)
+  merged = sample_examples(data, keep_classes, undersample)
+  write_tfexamples(merged, shards, output_dir)
+
+
+def read_tfexamples(p, uri, label):
+  """Function that reads tf.Examples from tfRecord files and converts them
+  to a K-V PCollection usable by Beam."""
+
+  dataset = tf.data.TFRecordDataset(tf.data.Dataset.list_files(f'{uri}/*'),
+                                    compression_type="GZIP")
+
+  # Take the input TFRecordDataset and extract the class label that we want.
+  # Output format is a K-V PCollection: {class_label: TFRecord in string format}
+  data = (p
+          | "DatasetToPCollection" >> beam.Create(dataset)
+          | "MapToLabel" >> beam.Map(generate_elements, label))
+  return data
+
+
+def sample_examples(data, keep_classes, undersample):
+  """Function that performs the sampling given a label-mapped dataset."""
+
+  # Finds the minimum frequency of all classes in the input label.
+  # Output is a singleton PCollection with the minimum # of examples.
+
+  def find_minimum(elements):
+    return min(elements or [0])
+
+  def find_maximum(elements):
+    return max(elements or [0])
+
+  sample_fn = find_minimum if undersample else find_maximum
```

Comment:
so if `undersample` flag is True, perform undersampling, otherwise perform oversampling? The name of this flag is not straightforward enough. Can we instead have something like `sampling_strategy` and enum values?
Please also clarify how to config this in the component doc.